storage: pebbleIterator panics too eagerly on Close()

After hooking the new pebbleIterator into the restore processor (https://github.com/cockroachdb/cockroach/pull/83984), a few restore roachtests on master have been failing (https://github.com/cockroachdb/cockroach/issues/84240, https://github.com/cockroachdb/cockroach/issues/84237), because the pebbleIterator panics when it attempts to close, for example: 
```
panic: EOF                                                                      
    9                                                                                 
   10 goroutine 45749 [running]:                                                      
   11 panic({0x483a900, 0xc0000a4120})                                                
   12         GOROOT/src/runtime/panic.go:1147 +0x3a8 fp=0xc0061cb528 sp=0xc0061cb468 pc=0x48a828
   13 github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).destroy(0xc0039f0000)
   14         github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:897 +0x1e5 fp=0xc0061cb668 sp=0xc0061cb528 pc=0x17a61a5
   15 github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).Close(0xc0039f0000)
```
(see full example panic stack trace in[ comment](https://github.com/cockroachdb/cockroach/issues/84240#issuecomment-1182279432)).

In the failed tests and in repros, the panic reasons have been `EOF`, `Context cancelled`, and `nil`, which at a high level, makes me wonder if all this panicking is necessary. To get a better understanding of when the iterator panics, I ran the `restore/nodeShutdown/coordinator` roachtest on my gce-worker using master + a [commit](https://github.com/cockroachdb/cockroach/commit/bf79431f9169f8765c7aa659e20e02fdd628f541) that adds more logging related to the SSTs in the iterator and issues prints instead of panics in `pebbleIterator.destroy()`. 

I can explain and reproduce the `context cancelled` panic. When the restore fails for some independent reason, context cancellations circulates throughout the goroutines, leading to pebbleIterator.Valid() returning `!ok`  and a `context cancelled` error. This then leads to `defer.Close()`, surfacing the `context cancelled` error during `destroy()`, leading to a panic. To repro this, stress the `restore/nodeShutdown/coordinator` roachtest above, potentially with my augmented commit.

I'm not sure what causes the EOF or nil based panics. 
@erikgrinaker 

Jira issue: CRDB-17633

Epic CRDB-2624

--- a/pkg/storage/pebble_iterator.go
+++ b/pkg/storage/pebble_iterator.go
@@ -892,10 +892,24 @@ func (p *pebbleIterator) destroy() {
 		panic("iterator still in use")
 	}
 	if p.iter != nil {
-		err := p.iter.Close()
-		if err != nil {
-			panic(err)
-		}
+		// If an error is encountered during iteration, it'll already have been
+		// surfaced by p.iter.Error() through Valid()'s error return value.
+		// Closing a pebble iterator that's in an error state surfaces that same
+		// error again. The client should've already handled the error when
+		// surfaced through Valid(), but wants to close the iterator (eg,
+		// potentially through a defer) and so we don't want to re-surface the
+		// error.
+		_ = p.iter.Close()
+
+		// TODO(jackson): In addition to errors accumulated during iteration,
+		// Close also returns errors encountered during the act of closing the
+		// iterator. Currently, these errors are swallowed. The error returned
+		// by iter.Close() may be an ephemeral error, or it may a misuse of the
+		// Iterator or corruption. Only swallow ephemeral errors (eg,
+		// DeadlineExceeded, etc), panic-ing on Close errors that are not known
+		// to be ephemeral/retriable.
+		//
+		// See cockroachdb/pebble#1811.
 		p.iter = nil
 	}
 	// Reset all fields except for the key and option buffers. Holding onto their

